src/shared/callbacks.rs

```diff
@@ -0,0 +1,147 @@
+use js_sys::{Function, Promise};
+use wasm_bindgen::prelude::*;
+use wasm_bindgen_futures::{spawn_local, JsFuture};
+
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll, Waker};
+
+#[derive(Default)]
+pub struct ResolvableState {
+    pub waker: Option<Waker>,
+    pub result: Vec<String>,
+}
+
+#[derive(Clone)]
+pub struct Resolvable {
+    pub resolved: Arc<Mutex<bool>>,
+    pub state: Arc<Mutex<ResolvableState>>,
+}
+
+#[allow(clippy::mutex_atomic)]
+impl Default for Resolvable {
+    fn default() -> Self {
+        Resolvable {
+            resolved: Arc::new(Mutex::new(false)),
+            state: Arc::new(Mutex::new(ResolvableState::default())),
+        }
+    }
+}
+
+impl Resolvable {
+    pub fn resolve(&mut self, val: Vec<String>) {
```

Comment:
Do we have to worry about a race condition here, since we're checking both mutexes at different times within the function?

Reply:
Ah, `lock()` blocks the current thread until it can access the mutex, that's cool.

Reply:
no because it is called inside the spawn local which resolves the future on the main thread